tests/integration/test_celery.py

```diff
@@ -0,0 +1,107 @@
+from pytest_docker_tools import container, fetch, wrappers
+
+from cmip_ref.database import Database
+from cmip_ref.datasets.cmip6 import CMIP6DatasetAdapter
+from cmip_ref.models import MetricExecutionResult
+from cmip_ref.solver import solve_metrics
+
+ROOT_DIR = Path(__file__).parents[2]
+
+
+class RedisContainer(wrappers.Container):
+    def ready(self):
+        if super().ready() and len(self.ports["6379/tcp"]) > 0:
+            print(f"Redis using port:{self.ports['6379/tcp'][0]}")
+            # Perform a simple ping to check if the server is ready
+            r = redis.Redis(host="localhost", port=self.ports["6379/tcp"][0])
+            try:
+                return r.ping()
+            except redis.ConnectionError:
+                return False
+
+        return False
+
+    def connection_url(self) -> str:
+        port = self.ports["6379/tcp"][0]
+        return f"redis://localhost:{port}/0"
+
+
+redis_image = fetch(repository="redis:7")
+
+redis_container = container(
+    image="{redis_image.id}",
+    ports={
+        "6379/tcp": None,
+    },
+    wrapper_class=RedisContainer,
+)
+
+
+@pytest.fixture
+def db_seeded(config, cmip6_data_catalog):
+    database = Database.from_config(config)
+
+    adapter = CMIP6DatasetAdapter()
+
+    with database.session.begin():
+        for instance_id, data_catalog_dataset in cmip6_data_catalog.groupby(adapter.slug_column):
+            adapter.register_dataset(config, database, data_catalog_dataset)
+
+    return database
+
+
+@pytest.fixture()
+def celery_app(redis_container, monkeypatch, config):
+    """
+    Fixture creating a Celery application instance.
+
+    This celery app is a bit different from the default one,
+    as it registers both to the "example" and "celery" queues.
+    Typically, these are done on separate workers.
+    """
+    monkeypatch.setenv("CELERY_BROKER_URL", redis_container.connection_url())
+    monkeypatch.setenv("CELERY_RESULT_BACKEND", redis_container.connection_url())
+
+    app = create_celery_app("test")
+
+    register_celery_tasks(app, provider)
+
+    import cmip_ref_celery.worker_tasks  # noqa
+
+    return app
+
+
+@pytest.fixture()
+def celery_worker_parameters():
+    return {"queues": ("example", "celery"), "perform_ping_check": False}
+
+
+def test_celery_solving(db_seeded, config, celery_worker, redis_container, monkeypatch):
+    config.executor.executor = "cmip_ref_celery.executor.CeleryExecutor"
+    monkeypatch.setenv("CELERY_BROKER_URL", redis_container.connection_url())
+    monkeypatch.setenv("CELERY_RESULT_BACKEND", redis_container.connection_url())
+
+    # Run the solver which executes the metrics
+    solve_metrics(db_seeded, timeout=10, config=config)
+
+    results = db_seeded.session.query(MetricExecutionResult).all()
+    assert len(results)
+    for result in results:
+        assert result.successful
+        assert (config.paths.results / result.output_fragment / result.path).exists()
+
+    gc.collect()
```

Comment:
@mikapfl This line took me hours. The `celery.AsyncResult`'s were being GCed after the redis container was torn down. This caused tests to hang sometimes depending on the order they were run. I went down a few different paths to try fix this before coalescing on this one line fix...

Celery testing is hard. We probably need an e2e test featuring docker containers etc, but that is a decent task in itself

Reply:
That sounds pretty painful. Async always is, it seems, but tests hanging because garbage collection doesn't happen wouldn't be on my list of things to expect either.